File: libs/ai-kernel/src/domain/agents/planner.rs
use crate::models::{
    agent::{AgentDefinition, AgentPlanStep},
    knowledge_base::KnowledgeSearchResult,
    tool::ToolDefinition,
};

pub fn build_plan(
    agent: &AgentDefinition,
    objective: &str,
    tools: &[ToolDefinition],
    knowledge_hits: &[KnowledgeSearchResult],
) -> Vec<AgentPlanStep> {
    let mut steps = vec![AgentPlanStep {
        id: "analyze-request".to_string(),
        title: "Analyze user intent".to_string(),
        description: format!("Align the request with agent objective '{}'.", objective),
        tool_name: None,
        status: "planned".to_string(),
    }];

    if !knowledge_hits.is_empty() {
        steps.push(AgentPlanStep {
            id: "retrieve-context".to_string(),
            title: "Retrieve supporting context".to_string(),
            description: format!(
                "Use {} retrieved passage(s) before drafting the answer.",
                knowledge_hits.len()
            ),
            tool_name: None,
            status: "planned".to_string(),
        });
    }

    for tool in tools.iter().take(agent.max_iterations.max(1) as usize) {
        steps.push(AgentPlanStep {
            id: format!("tool-{}", tool.name.to_lowercase().replace(' ', "-")),
            title: format!("Invoke {}", tool.name),
            description: tool.description.clone(),
            tool_name: Some(tool.name.clone()),
            status: "planned".to_string(),
        });
    }

    steps.push(AgentPlanStep {
        id: "synthesize-answer".to_string(),
        title: "Synthesize final answer".to_string(),
        description: format!(
            "Use planning strategy '{}' and produce an operator-friendly summary.",
            agent.planning_strategy
        ),
        tool_name: None,
        status: "planned".to_string(),
    });

    steps
}

File: openfoundry-go/libs/ai-kernel-go/domain/agents/planner.go
package agents

import (
	"fmt"
	"strings"

	"github.com/openfoundry/openfoundry-go/libs/ai-kernel-go/models"
)

// BuildPlan returns a deterministic plan-act-observe sequence.
// Mirrors Rust src/domain/agents/planner.rs verbatim:
//   1. analyze-request — always first.
//   2. retrieve-context — only when knowledge_hits is non-empty.
//   3. tool-{name} — first N tools where N = max(max_iterations, 1).
//   4. synthesize-answer — always last.
func BuildPlan(
	agent models.AgentDefinition,
	objective string,
	tools []models.ToolDefinition,
	knowledgeHits []models.KnowledgeSearchResult,
) []models.AgentPlanStep {
	steps := []models.AgentPlanStep{
		{
			ID:          "analyze-request",
			Title:       "Analyze user intent",
			Description: fmt.Sprintf("Align the request with agent objective '%s'.", objective),
			Status:      "planned",
		},
	}

	if len(knowledgeHits) > 0 {
		steps = append(steps, models.AgentPlanStep{
			ID:          "retrieve-context",
			Title:       "Retrieve supporting context",
			Description: fmt.Sprintf("Use %d retrieved passage(s) before drafting the answer.", len(knowledgeHits)),
			Status:      "planned",
		})
	}

	maxTools := int(agent.MaxIterations)
	if maxTools < 1 {
		maxTools = 1
	}
	if maxTools > len(tools) {
		maxTools = len(tools)
	}
	for i := 0; i < maxTools; i++ {
		t := tools[i]
		toolName := t.Name
		steps = append(steps, models.AgentPlanStep{
			ID:          fmt.Sprintf("tool-%s", strings.ReplaceAll(strings.ToLower(t.Name), " ", "-")),
			Title:       fmt.Sprintf("Invoke %s", t.Name),
			Description: t.Description,
			ToolName:    &toolName,
			Status:      "planned",
		})
	}

	steps = append(steps, models.AgentPlanStep{
		ID:          "synthesize-answer",
		Title:       "Synthesize final answer",
		Description: fmt.Sprintf("Use planning strategy '%s' and produce an operator-friendly summary.", agent.PlanningStrategy),
		Status:      "planned",
	})

	return steps
}
